src/clj/rems/ega.clj

```diff
@@ -0,0 +1,317 @@
+                      {:value "EGAD00001006673"
+                       :type "ControlledAccessGrants"
+                       :source "EGAC00001000908"
+                       :asserted 1615939200
+                       :by "dac"}
+                      :iat 1618617600}]]}
+           (update-in (entitlement->update {:resid "EGAD00001006673"
+                                            :userid "elixir-user"
+                                            :approvedby "dac-user"
+                                            :start (time-core/date-time 2021 03 17)
+                                            :end (time-core/date-time 2022 03 17)})
+                      [:visas 0]
+                      ga4gh/visa->claims)))))
+
+(defn entitlement-push [action entitlement config]
+  (let [api-key (get-api-key-for (:approvedby entitlement))
+        common-fields {:api-key api-key
+                       :account-id (:userid entitlement)
+                       :config config}]
+
+    (when (str/blank? api-key)
+      (log/warnf "Missing EGA api-key for %s" (:approvedby entitlement)))
+
+    (log/infof "Pushing entitlements to %s %s: %s %s" (getx config :id) (getx config :permission-server-url) entitlement config)
+
+    (case action
+      :add
+      (post-create-or-update-permissions (merge common-fields
+                                                (entitlement->update entitlement)))
+
+      :remove
+      (delete-permissions (merge common-fields
+                                 {:dataset-ids [(:resid entitlement)]})))))
+
+(defn generate-api-key
+  "Generates an API-Key and saves it to the user's secrets.
+
+  `:userid`               - REMS user to generate the key for
+  `:access-token`         - User's ELIXIR access token
+  `:config`               - configuration of the EGA integration with following keys:
+    `:connect-server-url` - EGA login server url
+    `:permission-server-url` - EGA permission server url
+    `:client-id`          - client id for REMS
+    `:client-secret`      - client secret for REMS"
+  [{:keys [userid access-token config]}]
+  (let [_ (log/info "Generate API-Key...")
+        expiration-date (time-core/plus (time-core/now) (time-core/years 1))
+        api-key (-> {:access-token access-token
+                     :id userid
+                     :expiration-date expiration-date
+                     :reason "rems_ega_push"
+                     :config config}
+                    get-api-key-generate
+                    :body
+                    :token)
+        _ (log/info "Save user secret...")
+        result (user-secrets/update-user-secrets! userid {:ega {:api-key api-key}})]
+    (if (:success result)
+      (do
+        (log/info "Success!")
+        {:success (:success result)
+         :api-key-expiration-date expiration-date})
+      {:success false})))
+
+(defn update-api-key
+  "Logs into EGA fetching an access token, generates an API-Key and saves it to the user's secrets.
```

Comment:
the distinction between generate and update seems a bit weird. perhaps something like `generate-api-key-with-access-token` and `generate-api-key-with-password`?

Reply:
Well generate does not sound like it should actually store it somewhere. I'll think of how to name these as services.

Reply:
I settled with `generate-api-key-with-access-token` and `generate-api-key-with-account` (as second one has username and password from EGA account).